presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/WindowPushdownRules.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.iterative.rule;
+
+import io.prestosql.metadata.FunctionId;
+import io.prestosql.sql.planner.PlanNodeIdAllocator;
+import io.prestosql.sql.planner.Symbol;
+import io.prestosql.sql.planner.plan.LimitNode;
+import io.prestosql.sql.planner.plan.TopNRowNumberNode;
+import io.prestosql.sql.planner.plan.WindowNode;
+
+import java.util.Optional;
+
+import static com.google.common.collect.Iterables.getOnlyElement;
+
+public class WindowPushdownRules
+{
+    private WindowPushdownRules()
+    {}
+
+    /**
+     * Check whether a window node can be target of the pushdown or replacement for optimization.
+     */
+    public static boolean canOptimizeWindowFunction(WindowNode node, FunctionId rowNumberFunctionId)
```

Comment:
This should be checked in the pattern of the rules that need this. BTW, the name is too vague. A better name would be `isRowNumber`